From af084551240270b344b27a5876b4f78811d7b408 Mon Sep 17 00:00:00 2001
From: Emmanuel Blot <emmanuel.blot@free.fr>
Date: Fri, 7 Aug 2026 16:27:25 +0200
Subject: [PATCH] hw/arm: bletchley: populate the v1.0 FRU EEPROMs

Back the eight FRU EEPROMs with images read from a v1.0 chassis, so
guest firmware can identify the six sled interposers, the storage
module and the chassis controller instead of finding them blank.

The content is sanitized: part and serial numbers masked, dates fixed,
asset tags replaced, distinct originals kept distinct.

Signed-off-by: Emmanuel Blot <emmanuel.blot@free.fr>
---
 hw/arm/aspeed_ast2600_bletchley.c | 309 +++++++++++++++++++++++++++++-
 1 file changed, 306 insertions(+), 3 deletions(-)

diff --git a/hw/arm/aspeed_ast2600_bletchley.c b/hw/arm/aspeed_ast2600_bletchley.c
index 7b006d0712..c3f7dfb27b 100644
--- a/hw/arm/aspeed_ast2600_bletchley.c
+++ b/hw/arm/aspeed_ast2600_bletchley.c
@@ -34,6 +34,304 @@
 #define BLETCHLEY_BMC_HW_STRAP2 0x00000801
 #define BLETCHLEY_BMC_RAM_SIZE ASPEED_RAM_SIZE(2 * GiB)
 
+/*
+ * FRU images read back from hardware and sanitized: part and serial
+ * numbers are masked, manufacturing dates fixed and asset tags replaced,
+ * so the layout stays faithful while the identifying data is not real.
+ */
+
+typedef struct BletchleyFru {
+    const uint8_t *data;
+    size_t size;
+} BletchleyFru;
+
+/* Every image a board revision needs, in i2c topology order. */
+typedef struct BletchleyFruSet {
+    BletchleyFru sled[6];   /* i2c0-5 @0x54 */
+    BletchleyFru bsm;       /* i2c6 @0x56 */
+    BletchleyFru cc;        /* i2c7 @0x54 */
+} BletchleyFruSet;
+
+/* Bletchley v1.0 */
+
+/*
+ * FRU: i2c0 @0x54 (sled 1 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000001
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_0_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x31, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xb3,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c1 @0x54 (sled 2 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000003
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_1_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x33, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xb1,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c2 @0x54 (sled 3 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000004
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_2_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x34, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xb0,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c3 @0x54 (sled 4 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000005
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_3_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x35, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xaf,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c4 @0x54 (sled 5 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000006
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_4_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x36, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xae,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c5 @0x54 (sled 6 interposer)
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Interposer-Class 1
+ *   Board part:      0AA00AA0001
+ *   Board serial:    AAA0A00000007
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_5_0054[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0c, 0x00, 0x00, 0xf2, 0x01, 0x0b, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xdc, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x49, 0x6e, 0x74, 0x65,
+    0x72, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73,
+    0x20, 0x31, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x37, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41,
+    0x30, 0x30, 0x30, 0x31, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72,
+    0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0xad,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c6 @0x56 (BMC storage module)
+ *   Board mfg:       Quanta
+ *   Board product:   BMC Storage Module
+ *   Board part:      0AA0AAA00A2
+ *   Board serial:    AAA0A00000008
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_6_0056[] = {
+    0x01, 0x00, 0x00, 0x01, 0x0b, 0x00, 0x00, 0xf3, 0x01, 0x0a, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xd2, 0x42, 0x4d,
+    0x43, 0x20, 0x53, 0x74, 0x6f, 0x72, 0x61, 0x67, 0x65, 0x20, 0x4d, 0x6f,
+    0x64, 0x75, 0x6c, 0x65, 0xcd, 0x41, 0x41, 0x41, 0x30, 0x41, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x30, 0x38, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x41,
+    0x41, 0x41, 0x30, 0x30, 0x41, 0x32, 0xcc, 0x46, 0x52, 0x55, 0x20, 0x56,
+    0x65, 0x72, 0x20, 0x30, 0x2e, 0x30, 0x34, 0xc1, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0xa2, 0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e,
+    0x74, 0x61, 0xcc, 0x42, 0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79,
+    0x20, 0x4d, 0x50, 0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30,
+    0x30, 0x30, 0x31, 0xc2, 0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45,
+    0x4d, 0x55, 0x00, 0xc1, 0x00, 0x00, 0x00, 0x99,
+};
+
+/*
+ * FRU: i2c7 @0x54 (chassis controller)
+ *   Chassis type:    Rack Mount Chassis
+ *   Chassis part:    0AA00AA0001
+ *   Chassis serial:  AAA0AA00AA000000A0A009A
+ *   Board mfg:       Quanta
+ *   Board product:   Bletchley Chassis Controller-Class 1
+ *   Board part:      0AA00AA0003
+ *   Board serial:    AAA0A00000010
+ *   Board mfg date:  2026-01-01T00:00:00Z
+ *   Board file ID:   FRU Ver 0.04
+ *   Board custom:    C01850000001
+ *   Product mfg:     Quanta
+ *   Product name:    Bletchley MP
+ *   Product part:    0AA00AA0001
+ *   Product version: MP
+ *   Product serial:  AAAA0000000002
+ *   Product asset:   QEMU
+ */
+static const uint8_t fru_v1_0_7_0054[] = {
+    0x01, 0x00, 0x01, 0x07, 0x15, 0x00, 0x00, 0xe2, 0x01, 0x06, 0x17, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xd7,
+    0x41, 0x41, 0x41, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x41, 0x30, 0x41, 0x30, 0x30, 0x39, 0x41, 0xc1,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x27, 0x01, 0x0e, 0x19, 0xc0,
+    0xc6, 0xf0, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xe4, 0x42, 0x6c,
+    0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x43, 0x68, 0x61, 0x73,
+    0x73, 0x69, 0x73, 0x20, 0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x6c,
+    0x65, 0x72, 0x2d, 0x43, 0x6c, 0x61, 0x73, 0x73, 0x20, 0x31, 0xcd, 0x41,
+    0x41, 0x41, 0x30, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x31, 0x30,
+    0xcb, 0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x33,
+    0xcc, 0x46, 0x52, 0x55, 0x20, 0x56, 0x65, 0x72, 0x20, 0x30, 0x2e, 0x30,
+    0x34, 0xc0, 0xc0, 0xcc, 0x43, 0x30, 0x31, 0x38, 0x35, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x31, 0xc1, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x11,
+    0x01, 0x08, 0x19, 0xc6, 0x51, 0x75, 0x61, 0x6e, 0x74, 0x61, 0xcc, 0x42,
+    0x6c, 0x65, 0x74, 0x63, 0x68, 0x6c, 0x65, 0x79, 0x20, 0x4d, 0x50, 0xcb,
+    0x30, 0x41, 0x41, 0x30, 0x30, 0x41, 0x41, 0x30, 0x30, 0x30, 0x31, 0xc2,
+    0x4d, 0x50, 0xce, 0x41, 0x41, 0x41, 0x41, 0x30, 0x30, 0x30, 0x30, 0x30,
+    0x30, 0x30, 0x30, 0x30, 0x32, 0xc4, 0x51, 0x45, 0x4d, 0x55, 0x00, 0xc1,
+    0x00, 0x00, 0x00, 0x99,
+};
+
+static const BletchleyFruSet bletchley_fru_v1_0 = {
+    .sled = {
+        { fru_v1_0_0_0054, sizeof(fru_v1_0_0_0054) },
+        { fru_v1_0_1_0054, sizeof(fru_v1_0_1_0054) },
+        { fru_v1_0_2_0054, sizeof(fru_v1_0_2_0054) },
+        { fru_v1_0_3_0054, sizeof(fru_v1_0_3_0054) },
+        { fru_v1_0_4_0054, sizeof(fru_v1_0_4_0054) },
+        { fru_v1_0_5_0054, sizeof(fru_v1_0_5_0054) },
+    },
+    .bsm = { fru_v1_0_6_0056, sizeof(fru_v1_0_6_0056) },
+    .cc  = { fru_v1_0_7_0054, sizeof(fru_v1_0_7_0054) },
+};
+
 #define TYPE_BLETCHLEY_MACHINE MACHINE_TYPE_NAME("bletchley-bmc")
 OBJECT_DECLARE_SIMPLE_TYPE(BletchleyMachineState, BLETCHLEY_MACHINE)
 
@@ -45,6 +343,7 @@ struct BletchleyMachineState {
 
 static void bletchley_bmc_i2c_init(AspeedMachineState *bmc)
 {
+    const BletchleyFruSet *fru = &bletchley_fru_v1_0;
     AspeedSoCState *soc = bmc->soc;
     I2CBus *i2c[13] = {};
     for (int i = 0; i < 13; i++) {
@@ -63,17 +362,21 @@ static void bletchley_bmc_i2c_init(AspeedMachineState *bmc)
         i2c_slave_create_simple(i2c[i], TYPE_PCA9539, 0x76);
         i2c_slave_create_simple(i2c[i], TYPE_PCA9552, 0x67);
         /* Missing model: fsc,fusb302 @ 0x22 */
-        at24c_eeprom_init(i2c[i], 0x54, 8 * KiB);
+        at24c_eeprom_init_rom(i2c[i], 0x54, 8 * KiB,
+                              fru->sled[i].data,
+                              fru->sled[i].size);
     }
 
     /* Bus 6 */
-    at24c_eeprom_init(i2c[6], 0x56, 8 * KiB);
+    at24c_eeprom_init_rom(i2c[6], 0x56, 8 * KiB,
+                          fru->bsm.data, fru->bsm.size);
     /* Missing model: nxp,pcf85263 @ 0x51 , but ds1338 works enough */
     i2c_slave_create_simple(i2c[6], TYPE_DS1338, 0x51);
 
 
     /* Bus 7 */
-    at24c_eeprom_init(i2c[7], 0x54, 8 * KiB);
+    at24c_eeprom_init_rom(i2c[7], 0x54, 8 * KiB,
+                          fru->cc.data, fru->cc.size);
 
     /* Bus 9 */
     i2c_slave_create_simple(i2c[9], TYPE_TMP421, 0x4f);
-- 
2.53.0-Meta

